js/bitopro.js

```diff
@@ -0,0 +1,1349 @@
+        //
+        //     {
+        //         "bids":[
+        //             {
+        //                 "price":"1175271",
+        //                 "amount":"0.00022804",
+        //                 "count":1,
+        //                 "total":"0.00022804"
+        //             }
+        //         ],
+        //         "asks":[
+        //             {
+        //                 "price":"1176906",
+        //                 "amount":"0.0496",
+        //                 "count":1,
+        //                 "total":"0.0496"
+        //             }
+        //         ]
+        //     }
+        //
+        return this.parseOrderBook (response, symbol, timestamp, 'bids', 'asks', 'price', 'amount');
+    }
+
+    parseTrade (trade, market) {
+        //
+        // fetchTrades
+        //         {
+        //             "timestamp":1644651458,
+        //             "price":"1180785.00000000",
+        //             "amount":"0.00020000",
+        //             "isBuyer":false
+        //         }
+        //
+        // fetchMyTrades
+        //         {
+        //             "id":"8384438099",
+        //             "pair":"bnb_twd",
+        //             "price":"16000",
+        //             "avgExecutionPrice":"0",
+        //             "action":"SELL",
+        //             "type":"LIMIT",
+        //             "timestamp":1644902118265,
+        //             "status":4,
+        //             "originalAmount":"0.01",
+        //             "remainingAmount":"0.01",
+        //             "executedAmount":"0",
+        //             "fee":"0",
+        //             "feeSymbol":"twd",
+        //             "bitoFee":"0",
+        //             "total":"0",
+        //             "seq":"BNBTWD8984230567",
+        //             "updatedTimestamp":1644902118265
+        //         }
+        //
+        const id = this.safeString (trade, 'id');
+        let timestamp = this.safeInteger (trade, 'timestamp');
+        if (id === undefined) {
+            timestamp = this.safeTimestamp (trade, 'timestamp');
+        }
+        const marketId = this.safeString (trade, 'pair');
+        market = this.safeMarket (marketId, market);
+        const symbol = this.safeString (market, 'symbol');
+        const price = this.safeString (trade, 'price');
+        let type = this.safeString (trade, 'type');
+        if (type !== undefined) {
+            type = type.toLowerCase ();
+        }
+        let side = this.safeString (trade, 'action');
+        if (side === undefined) {
+            const isBuyer = this.safeValue (trade, 'isBuyer');
+            if (isBuyer) {
+                side = 'buy';
+            } else {
+                side = 'sell';
+            }
+        } else {
+            side = side.toLowerCase ();
+        }
+        let amount = this.safeString (trade, 'amount');
+        if (amount === undefined) {
+            amount = this.safeFloat (trade, 'executedAmount');
```

Comment:
`safeFloat` is deprecated, please use `safeNumber` instead

Reply:
Should I use safeString here?

Reply:
We should use `safeString2` in such cases.